### PR TITLE
Add statistics to TransactionView

### PR DESCRIPTION
**Pull request by wasanderes94**
### Context
For code-week we should implement a statistics feature, which displays stuff like most bought items or balance over time

- Added chart library to gradle file and used `AndroidView` to use this library with Jetpack Compose
- Added PieChart for category statistics. Styled Chart and Legend
- Added `getPurchaseOfUser` method to `TransactionViewModel` which is fetching purchases from DB
- Added `getFundingOfUser` method to `TransactionViewModel` which is fetching transactions from API, filtering only for Fudings and, if there are more fundings received than contained in DB, adding them.
- Added `getBalanceOfUser` method to `TransactionViewModel` which is creating a custom data structure for handling the balance inside the LineChart
- Implemented LineChart for balance statistics
- Formatted x-Values of LineChart to obtain a daily scale instead of timestamps
- Made both charts updatable
- Committed a basic implementation for a background worker. This was intended to be used for Chart Updates but is not useful, since it only can pull values at an interval of 15 Minutes (see #115 )

### Known issues
The last value of lineChart hidden behind view borders and therefore not displayed. Might investigate this.
The PieChart has a legend for the 4 most consumed items but is displaying every consumed item. This leads to slices which share the same colour but one does not have any legend or Entry label

### CODE REVIEW

**Review by SteffenKr — COMMENTED** [review]
Fix the TODO's and add string resources when needed

**Review by Morphclue — CHANGES_REQUESTED** [review]
You can remove a lot of comments inside the code, because you are often times only describing what the next line does and in these cases the lines describe themself.

eg. : 
```kt
// fetch transactions from API
val response = user.repository.getTransactions(...)
```
If you want to logically cluster parts of a method you can refactor it into multiple new methods.
e.g.
```kt
// circle styling
// - Line 1
// - Line 2
// legend styling
// - Line 1
// - Line 2
// - Line 3
``` could be refactored into
```kt
stylePieChart()
styleLegends()
```

File: app/src/main/java/org/feature/fox/coffee_counter/ui/transaction/TransactionView.kt [hunk 1]
```diff
@@ -114,6 +131,133 @@ fun QRCodeButton(viewModel: ITransactionViewModel) {
...
+                pieChart.setNoDataText("No Purchases found")
+
+                pieChart
+            },
+            update = { pieChart ->
+                val listColors = ArrayList<Int>()
+                listColors.add(Color(52, 152, 219, 255).toArgb()) //Blue
+                listColors.add(Color(230, 76, 59, 255).toArgb()) //Red
+                listColors.add(Color(241, 196, 15, 255).toArgb()) //Yellow
+                listColors.add(Color(46, 204, 112, 255).toArgb()) //Green
```
[the remark this answers is not in the document]
Comment (wasanderes94) [reply app/src/main/java/org/feature/fox/coffee_counter/ui/transaction/TransactionView.kt]: Should be resolved with e619d0dbf4914c496c72810a38bcc1bf783bc547

File: app/build.gradle [hunk 2]
```diff
@@ -146,6 +152,19 @@ dependencies {
...
+    implementation "com.github.PhilJay:MPAndroidChart:$mpandroidchart_version"
+
+
+    // Work Manager
+    implementation "androidx.work:work-runtime-ktx:$work_version"
+    implementation "androidx.work:work-gcm:$work_version"
+    implementation "androidx.work:work-multiprocess:$work_version"
+    // Hilt Workmanager
+    implementation 'androidx.hilt:hilt-work:1.0.0'
+    kapt 'androidx.hilt:hilt-compiler:1.0.0'
```
[the remark this answers is not in the document]
Comment (wasanderes94) [reply app/build.gradle]: Should be resolved with da4db111c4281676f48b0c3972fa15b976d02fb2

File: app/src/main/java/org/feature/fox/coffee_counter/data/local/database/dao/UserDao.kt [hunk 3]
```diff
@@ -41,4 +43,13 @@ interface UserDao {
...
     @Delete
     suspend fun deleteImageDb(image: Image)
+
+    @Transaction
+    @Query("SELECT * FROM purchase WHERE userId = :userId")
+    suspend fun getPurchasesOfUserByIdDb(userId: String): List<Purchase>
+
+    @Transaction
+    @Query("SELECT * FROM funding WHERE userId = :userId")
+    suspend fun getFundingOfUserByIdDb(userId: String): List<Funding>
```
[the remark this answers is not in the document]
Comment (wasanderes94) [reply app/src/main/java/org/feature/fox/coffee_counter/data/local/database/dao/UserDao.kt:47]: Should be resolved with f0b3955fa9c3e9435e37324bd8edeb2b97046f0e

File: app/src/main/java/org/feature/fox/coffee_counter/data/repository/UserRepositoryInt.kt [hunk 4]
```diff
@@ -33,6 +33,10 @@ interface UserRepositoryInt {
 
     suspend fun deleteUserDb(user: User)
 
+    suspend fun getPurchasesOfUserByIdDb(userId: String): List<Purchase>
+
+    suspend fun getFundingOfUserByIdDb(userId: String): List<Funding>
```
[the remark this answers is not in the document]
Comment (wasanderes94) [reply app/src/main/java/org/feature/fox/coffee_counter/data/repository/UserRepositoryInt.kt:36]: Should be resolved with f0b3955fa9c3e9435e37324bd8edeb2b97046f0e

File: app/src/main/java/org/feature/fox/coffee_counter/ui/transaction/TransactionView.kt [hunk 5]
```diff
@@ -114,6 +131,133 @@ fun QRCodeButton(viewModel: ITransactionViewModel) {
...
+                val lineChart = LineChart(context)
+                val formatter = DateTimeFormatter()
+                lineChart.xAxis.position = XAxis.XAxisPosition.BOTTOM
+                lineChart.xAxis.setDrawGridLines(false)
+                lineChart.xAxis.granularity = 2f
+                lineChart.xAxis.valueFormatter = formatter
+                lineChart.axisLeft.setDrawGridLines(false)
+                lineChart.axisRight.isEnabled = false
+                lineChart.setNoDataText("No funding/orders yet")
+                lineChart.description.text = "Balance over Time"
```
[the remark this answers is not in the document]
Comment (wasanderes94) [reply app/src/main/java/org/feature/fox/coffee_counter/ui/transaction/TransactionView.kt]: Should be resolved with c3e1186232193b1e768d17281adb1da00b8d8da5

**Comment by wasanderes94** [comment]
> Missing new line.

Should be resolved with 086f3ab36735f2cda6695c2e6f98e33c47657acb

**Comment by wasanderes94** [comment]
> You can remove a lot of comments inside the code, because you are often times only describing what the next line does and in these cases the lines describe themself.
> 
> eg. :
> 
> ```kotlin
> // fetch transactions from API
> val response = user.repository.getTransactions(...)
> ```
> 
> If you want to logically cluster parts of a method you can refactor it into multiple new methods. e.g.
> 
> ```kotlin
> // circle styling
> // - Line 1
> // - Line 2
> // legend styling
> // - Line 1
> // - Line 2
> // - Line 3
> ```
> 
> could be refactored into
> 
> ```kotlin
> stylePieChart()
> styleLegends()
> ```

Should be resolved with c3e1186232193b1e768d17281adb1da00b8d8da5 although i am a fan of commenting code :P

**Review by Morphclue — APPROVED** [review]
This should be fine after moving the `Share money via QR-Code` as a sticky on top.

**Comment by wasanderes94** [comment]
Moved `Share money via QR-Code` code on top with 2829be5989327255feb1a99b1c4db0995ec54bde

**Pull request merged**